File: yggy/observable/observable_network.ts
import { Comm } from "../comm/comm.js";
import {
    ChangeMessage,
    OBSERVABLE_CHANGE_MSG,
    OBSERVABLE_CLIENT_CHANGE_MSG,
} from "./messages.js";
import { Observable } from "./observable.js";

export class ObservableNetwork {
    #registry: { [key: string]: Observable<any> };
    #updating: Set<string>;
    #comm: Comm;

    constructor(__comm: Comm) {
        this.#registry = {};
        this.#updating = new Set();
        this.#comm = __comm;

        this.#comm.recv(OBSERVABLE_CHANGE_MSG, this.#recv_change.bind(this));
    }

    public get comm(): Comm {
        return this.#comm;
    }

    public get(__id: string | number): Observable<any> | undefined {
        if (typeof __id === "number") {
            return Object.values(this.#registry)[__id];
        }
        return this.#registry[__id];
    }

    public send_change(__change: ChangeMessage<any>): void {
        if (this.#updating.has(__change.data_id)) {
            return;
        }

        this.#comm.send(OBSERVABLE_CLIENT_CHANGE_MSG, __change);
    }

    public notify_change(__change: ChangeMessage<any>): void {
        this.#comm.notify(OBSERVABLE_CHANGE_MSG, __change);
    }

    public register(__obs: Observable<any>): void {
        this.#registry[__obs.id] = __obs;
        __obs.register(this);
    }

    #recv_change(__change: ChangeMessage<any>): void {
        const { data_id } = __change;

        if (this.#updating.has(data_id)) {
            return;
        }
        this.#updating.add(data_id);

        this.#registry[data_id].__recv_change(__change);

        this.#updating.delete(data_id);
    }

}

File: yggy/observable/observable_network.py
from typing import Any, Iterable
from weakref import WeakKeyDictionary, WeakValueDictionary

from ..comm import Comm, LazyMessage
from .messages import OBSERVABLE_CHANGE_MSG, OBSERVABLE_CLIENT_CHANGE_MSG, ChangeMessage
from .observable import Observable

__all__ = [
    "ObservableNetwork",
]


class ObservableNetwork:
    __registry: WeakValueDictionary[str, Observable[Any]]
    __client_registry: WeakKeyDictionary[Observable[Any], set[str]]
    __comm: Comm

    def __init__(self, __comm: Comm) -> None:
        self.__registry = WeakValueDictionary()
        self.__client_registry = WeakKeyDictionary()
        self.__comm = __comm

        self.__comm.recv(OBSERVABLE_CLIENT_CHANGE_MSG, self.__recv_client_change)

    @property
    def comm(self) -> Comm:
        return self.__comm

    def register(
        self,
        __observables: Observable[Any] | Iterable[Observable[Any]],
        clients: Iterable[str] | None = None,
    ) -> None:
        if isinstance(__observables, Iterable):
            observables = list(__observables)
        else:
            observables = [__observables]

        for observable in observables:
            self.__registry[observable.id] = observable
            observable._register(self)  # type: ignore[friends til the end]

        if clients is None:
            return

        for observable in observables:
            cur_clients = self.__client_registry.setdefault(observable, set[str]())
            cur_clients.update(clients)

    # def unregister(self, __observable: Observable[Any]) -> None:
    #     del self.__registry[__observable.id]

    def emit_change(
        self,
        __observable: Observable[Any],
        __change: ChangeMessage[Any] | LazyMessage[ChangeMessage[Any]],
        *,
        revoke: str | None = None,
    ) -> None:
        if revoke is not None:
            self.__comm.revoke(revoke)
        self.__comm.emit(
            OBSERVABLE_CHANGE_MSG,
            __change,
            client_ids=self.__client_registry.get(__observable, []),
        )

    def __recv_client_change(self, __change: ChangeMessage[Any]) -> None:
        data_id = __change["data_id"]

        if data_id in self.__registry:
            obs = self.__registry[data_id]

            new_value = __change["new_value"]
            obs.set(new_value)
